rebalance/rebalance.py

```diff
@@ -212,16 +215,260 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi=None,
+                i += 1
+            if amount > get_max_amount(i, min_amount, plugin.enough_liquidity):
+                continue
+            return result
+        plugin.log(f"Rebalance succeeded: {res}")
+        result["success"] = True
+        result["fee_spent"] += res["fee"]
+        # refresh channels
+        time.sleep(10)
+        ch1 = get_chan(plugin, scid1)
+        assert ch1 is not None
+        ch2 = get_chan(plugin, scid2)
+        assert ch2 is not None
+
+
+def maybe_rebalance_once(plugin: Plugin, failed_pairs: list):
+    channels = get_our_channels(plugin)
+    for ch1 in channels:
+        for ch2 in channels:
+            result = maybe_rebalance_pairs(plugin, ch1, ch2, failed_pairs)
+            if result["success"] or plugin.rebalance_stop:
+                return result
+    return {"success": False, "fee_spent": 0}
+
+
+def rebalanceall_thread(plugin: Plugin):
+    try:
+        channels = get_our_channels(plugin)
+        plugin.enough_liquidity = get_enough_liquidity_threshold(channels)
+        plugin.ideal_ratio = get_ideal_ratio(channels, plugin.enough_liquidity)
+        plugin.log(f"Automatic rebalance is running with enough liquidity threshold: {plugin.enough_liquidity:,} msat, "
+                   f"ideal liquidity ratio: {plugin.ideal_ratio * 100:.2f}%, "
+                   f"min rebalancable amount: {int(plugin.min_amount):,} msat, "
+                   f"feeratio: {plugin.feeratio}")
+        failed_pairs = []
+        success = 0
+        fee_spent = 0
+        while not plugin.rebalance_stop:
+            result = maybe_rebalance_once(plugin, failed_pairs)
+            if not result["success"]:
+                break
+            success += 1
+            fee_spent += result["fee_spent"]
+            feeadjust_would_be_nice(plugin)
+        plugin.log(f"Automatic rebalance finished: {success} successful rebalance, {fee_spent:,} msat fee spent")
+    finally:
+        plugin.mutex.release()
+
+
+@plugin.method("rebalanceall")
+def rebalanceall(plugin: Plugin, min_amount: Millisatoshi = Millisatoshi("50000sat"), feeratio: float = 0.5):
+    """Rebalance all unbalanced channels if possible for a very low fee.
+    Default minimum rebalancable amount is 50000sat. Default feeratio = 0.5, half of our node's default fee.
+    To be economical, it tries to fix the liquidity cheaper than it can be ruined by transaction forwards.
+    It may run for a long time (hours) in the background, but can be stopped with the rebalancestop method.
+    """
+    if not plugin.mutex.acquire(blocking=False):
+        return {"message": "Rebalance is already running, this may take a while. To stop it use the cli method 'rebalancestop'."}
+    try:
+        plugin.feeratio = float(feeratio)
+        plugin.min_amount = Millisatoshi(min_amount)
+        t = Thread(target=rebalanceall_thread, args=(plugin, ))
+        t.start()
+    except Exception as e:
+        plugin.mutex.release()
```

Comment:
When the `mutex` is `acquire` within the thread run function, this `release` does not need to be here... but in the threads generic exception/finally handler (where it already is :)